merge ReinitDone into Init

ulmus-scott opened the pull request:
I'm not sure why it was ever split, and it is currently only called immediately
after Init.
- Removed duplicated calls from ReinitDone
- Improved logic for recreating painters

Thank you for contributing to MythTV!

Please review the checklist below to ensure that your contribution
to MythTV is ready for review by our developers.

It is helpful to regularly rebase your pull request to ensure changes
apply cleanly to the current MythTV development branch.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] contribution is in a branch rebased against [master](https://github.com/MythTV/mythtv)
- [x] code compiles successfully without errors
- [x] code follows the [MythTV Coding Standards](https://www.mythtv.org/wiki/Coding_Standards)
- [x] commits are logically organised and have [good commit messages](https://chris.beams.io/posts/git-commit)

kmdewaal commented:
Please state which problem this pull request is solving.

For the information, the ReInitDone function is introduced in 2009 in commit ea71db0db79a2c13d48990a1067ac5acc25687c7 "Switching themes/resolutions/painters on the fly now works." From no other than Isaac Richards himself.

I have tested this patch with today's master and mythtv-setup.
Unfortunately, mythtv-setup crashes before the first window is drawn:
```
2021-08-05 17:47:53.850340 I  Display: Have main window
2021-08-05 17:47:53.850369 I  Display: Found screen 'eDP-1'
2021-08-05 17:47:53.851697 E  MythMainWindow failed to create a painter window.
2021-08-05 17:47:53.857254 I  Current MythTV Schema Version (DBSchemaVer): 1370
[New Thread 0x7fff8d687640 (LWP 192056)]
[New Thread 0x7fff8ce86640 (LWP 192057)]
[New Thread 0x7fff77fff640 (LWP 192058)]
[New Thread 0x7fff777fe640 (LWP 192059)]
[New Thread 0x7fff76ffd640 (LWP 192060)]
[New Thread 0x7fff767fc640 (LWP 192061)]
[New Thread 0x7fff75ffb640 (LWP 192062)]
[New Thread 0x7fff757fa640 (LWP 192063)]
2021-08-05 17:47:53.927776 E  MythFontProperties: Specified base font 'small' does not exist.
                        Location: /home/klaas/mma-dev/usr/share/mythtv/themes/MythCenter-wide/menu-ui.xml @ 17
                        Name: 'clock'   Type: 'fontdef'

Thread 29 "PT4" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff76ffd640 (LWP 192060)]
QMutex::lock (this=0x28) at thread/qmutex.cpp:230
230         if (fastTryLock(current))
Missing separate debuginfos, use: dnf debuginfo-install libcurl-7.76.1-4.fc34.x86_64 sssd-client-2.5.1-2.fc34.x86_64 systemd-libs-248.4-1.fc34.x86_64
(gdb) bt
#0  QMutex::lock() (this=0x28) at thread/qmutex.cpp:230
#1  0x00007ffff6bf28e1 in QMutexLocker::QMutexLocker(QBasicMutex*) (this=0x7fff76ffc5b8, m=0x28) at /usr/include/qt5/QtCore/qmutex.h:233
#2  0x00007ffff6c01629 in MythPainter::GetFormatImage() (this=0x0) at mythpainter.cpp:542
#3  0x00007ffff6c49416 in ImageLoader::LoadImage(MythPainter*, ImageProperties, ImageCacheMode, MythUIImage const*, bool&, MythImageReader*)
    (painter=0x0, imProps=..., cacheMode=kCacheNormal, parent=0xbad7e0, aborted=@0x7fff76ffc7ff: false, imageReader=0x0) at mythuiimage.cpp:274
#4  0x00007ffff6c4a784 in ImageLoadThread::run() (this=0xbade00) at mythuiimage.cpp:521
#5  0x00007ffff6ebfec1 in MPoolThread::run() (this=0xbadec0) at mthreadpool.cpp:142
#6  0x00007ffff6ebc390 in MThreadInternal::run() (this=0xbadf20) at mthread.cpp:78
#7  0x00007ffff555a456 in QThreadPrivate::start(void*) (arg=0xbadf20) at thread/qthread_unix.cpp:329
#8  0x00007ffff694e299 in start_thread (arg=0x7fff76ffd640) at pthread_create.c:481
#9  0x00007ffff5166353 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb)

Looks to me that the split in Init and ReInitDone does have a function.
Note that the behavior can be dependent on the environment; this was tested with Fedora 34, Wayland and Intel graphics.

ulmus-scott commented:
> Please state which problem this pull request is solving.

No problem, just code cleanup that I noticed could be done while investigating https://forum.mythtv.org/viewtopic.php?p=22281#p22285

> I have tested this patch with today's master and mythtv-setup.
Unfortunately, mythtv-setup crashes before the first window is drawn:

I tested it again.  I didn't get a crash, but it failed to open.

I fixed the logic for (re)creating the painters, since as I had changed it, it only recreated the painters, but didn't create them if they didn't exist.

I also separated that change from merging ReinitDone.

> For the information, the ReInitDone function is introduced in 2009 in commit ea71db0 "Switching themes/resolutions/painters on the fly now works." From no other than Isaac Richards himself.

I know, but the splitting is, in my opinion, a hack-y and inelegant way to insert a few lines for the frontend.  Regardless, since it is no longer called with intervening lines, it no longer serves any purpose.  If necessary, I think the (duplicated) calls from ReinitDone could safely be added at the end of Init.

ulmus-scott commented:
The reason the split is hack-y and bad, is that it makes bugs more likely and hard to find if you only call Init **because it is not obvious you also need to always call ReinitDone** if HasMythMainWindow() is true.  You could check for that, but ReinitDone is not necessary anymore (in master).

As master stands right now, not calling ReinitDone should only leak memory (if I interpret the code correctly).

In v31:
If m_paintwin exists, the draw timer is stopped, but only restarted in ReinitDone.
https://github.com/MythTV/mythtv/blob/fixes/31/mythtv/libs/libmythui/mythmainwindow.cpp#L1081
https://github.com/MythTV/mythtv/blob/fixes/31/mythtv/libs/libmythui/mythmainwindow.cpp#L1360

Only calls Init:
https://github.com/MythTV/mythtv/blob/master/mythtv/programs/mythwelcome/main.cpp#L120
https://github.com/MythTV/mythtv/blob/master/mythtv/programs/mythavtest/main.cpp#L249
https://github.com/MythTV/mythtv/blob/master/mythtv/programs/mythfrontend/main.cpp#L2015 also notice inconsistent use of GetMythMainWindow() i.e. multiple variables storing that elsewhere in main()
https://github.com/MythTV/mythtv/blob/master/mythtv/programs/mythscreenwizard/main.cpp#L180

https://github.com/MythTV/mythtv/blob/master/mythtv/libs/libmyth/mythcontext.cpp#L291 should be fine, however, because it is only called if a mainWindow does not exist


root cause of https://forum.mythtv.org/viewtopic.php?f=36&t=4470&start=45#p22285 :
https://github.com/MythTV/mythtv/blob/master/mythtv/programs/mythtv-setup/main.cpp#L379 create a MythContext
then MythContext::Init to
MythContextPrivate::Init to
MythContextPrivate::FindDatabase to
MythContextPrivate::TestDBconnection where ShowGuiStartup is called if msStartupScreenDelay==0ms
ShowGuiStartup then calls TempMainWindow, which creates a MythMainWindow before
https://github.com/MythTV/mythtv/blob/master/mythtv/programs/mythtv-setup/main.cpp#L544 Init without ReinitDone as I suspected

I'm not that familiar with git or github, but only the last two commits are necessary (the others are either extraneous or reverted).

kmdewaal commented on mythtv/libs/libmythui/mythmainwindow.cpp:
```diff
@@ -715,22 +715,18 @@ void MythMainWindow::Init(bool MayReInit)
-    if (!m_painter && !m_painterWin)
-        warningmsg = MythPainterWindow::CreatePainters(this, m_painterWin, m_painter);
-
+    
```
White space, please remove

ulmus-scott replied:
Remove just the spaces or the blank line?

ulmus-scott replied:
diff incorrectly interpreted the braces as not being changed
it should show remove:

```
if (m_painter)
{
    m_oldPainter = m_painter;
    m_painter = nullptr;
}
```

modify line 730
remove 731 and 732
add

```
if (!warningmsg.isEmpty())
{
    LOG(VB_GENERAL, LOG_WARNING, warningmsg);
}
```

Is a single line if statement considered acceptable style?  i.e.:

```
if (!warningmsg.isEmpty()) LOG(VB_GENERAL, LOG_WARNING, warningmsg);
```

kmdewaal replied:
Single line if statements are generally never used. A single statement on the next line is accepted by the code checkers but only if the statement fits on a single line. Concluding, the style with the { } brackets even for a single following statement is preferred.
There is information in the Wiki https://www.mythtv.org/wiki/Coding_Standards but the actual code checkers are much much more strict.

kmdewaal replied:
> Remove just the spaces or the blank line?

The spaces is what git complains about when creating a patch of your pull request and then applying that patch.

kmdewaal commented on mythtv/libs/libmythui/mythmainwindow.cpp:
```diff
@@ -775,6 +771,10 @@ void MythMainWindow::Init(bool MayReInit)
         MythNotification notification(warningmsg, "");
         m_priv->m_nc->Queue(notification);
     }
+    
```
I think these four lines can go, if the calls are not needed anymore then they do not need to remain as comment.

kmdewaal replied:
I did intend to mark these lines for removal:
    
    // from ReinitDone
    //ShowPainterWindow();
    //MoveResize(m_screenRect);

ulmus-scott commented:
I removed the spaces on the empty lines.

I thought it might be useful for reference to have the comments with the duplicated calls if their removal had some unexpected side effect.  However, I agree, they shouldn't be necessary, so are now removed.

The relevant commits are:
https://github.com/MythTV/mythtv/pull/369/commits/595ba00400223959a5269fff1963342425d0388d
https://github.com/MythTV/mythtv/pull/369/commits/5a79f5e94ee34b2c75c237be6355a892d31b6c78
https://github.com/MythTV/mythtv/pull/369/commits/226320d906a3d9175616f98677c5c72f06ce1c9b
